aiokafka/message_accumulator.py

```diff
@@ -0,0 +1,185 @@
+            elif base_offset is None:
+                future.set_result(None)
+            else:
+                res = RecordMetadata(self._tp.topic, self._tp.partition,
+                                     self._tp, base_offset+relative_offset)
+                future.set_result(res)
+
+    @asyncio.coroutine
+    def wait_all(self):
+        """wait until all message from this batch is processed"""
+        yield from asyncio.wait(self._msg_futures, loop=self._loop)
+
+    def expired(self):
+        """check that batch is expired or not"""
+        return (self._loop.time() - self._ctime) > self._ttl
+
+    def pack(self):
+        """close batch to be ready for send"""
+        self._records.close()
+
+    def data(self):
+        return self._records.buffer()
+
+
+class MessageAccumulator:
+    """Accumulator of messages batches by topic-partition
+
+    Producer add messages to this accumulator and background send task
+    gets batches per nodes for process it.
+    """
+    def __init__(self, cluster, batch_size, compression_type, ttl, loop):
+        self._batches = {}
+        self._cluster = cluster
+        self._batch_size = batch_size
+        self._compression_type = compression_type
+        self._ttl = ttl
+        self._loop = loop
+        self._wait_data_future = asyncio.Future(loop=loop)
+        self._empty_futures = {}
+        self._closed = False
+
+    def close(self):
+        self._closed = True
+
+    @asyncio.coroutine
+    def add_message(self, tp, key, value):
+        """Add message to batch by topic-partition
+        If batch is already full this method waits (`ttl` seconds maximum)
+        until batch is drained by send task
+        """
+        if self._closed:
+            # this can happen when producer is closing but try to send some
+            # messages in async task
+            raise ProducerClosed()
+
+        batch = self._batches.get(tp)
+        if not batch:
+            message_set_buffer = MessageSetBuffer(
+                io.BytesIO(), self._batch_size, self._compression_type)
+            batch = MessageBatch(tp, message_set_buffer, self._ttl, self._loop)
+            self._batches[tp] = batch
+
+            efut = self._empty_futures.get(tp)
+            if efut is None or efut.done():
+                # batch per topic-partition is not empty
+                # so create new "empty" future
+                self._empty_futures[tp] = asyncio.Future(loop=self._loop)
+            if not self._wait_data_future.done():
+                # we have some data, so resolve "wait data" future
+                self._wait_data_future.set_result(None)
+
+        future = batch.append(key, value)
+        if future is None:
+            # waiting until batch per topic-partition is drained
+            done, _ = yield from asyncio.wait(
+                [self._empty_futures[tp]], timeout=self._ttl, loop=self._loop)
+            if not done:
+                raise KafkaTimeoutError()
```

Comment:
Am I correct to assume, that this error will be only if we have so much messages, that we 
* Filled the first batch, sent it
* Filled the second batch
* And could not send the first one for over `ttl` seconds